### PR TITLE
ENH Adds nn.ReflectionPad3d

### LINKED ISSUE #27655: Why we do not have torch.nn.ReflectionPad3d?
We have used ReflectionPad2d for cyclegan and it is very important to solve the checkboard artifact. 

Considering the 5D case, I also meet the checkboard artifact with cyclegan. I try to find the ReflectionPad3d with the newest PyTorch version but it still in the checklist. Could you tell me when the function will available? It is very useful if it can integrate it in PyTorch

### PR DESCRIPTION
**Pull request by thomasjpfan**
Fixes https://github.com/pytorch/pytorch/issues/27655

This PR adds a C++ and Python version of ReflectionPad3d with structured kernels. The implementation uses lambdas extensively to better share code from the backward and forward pass.

### CODE REVIEW

File: test/cpp/api/modules.cpp [hunk 1]
```diff
@@ -3987,6 +3987,53 @@ TEST_F(ModulesTest, ReflectionPad2d) {
...
+                                      {1., 0., 1., 0.},
+                                      {3., 2., 3., 2.},
+                                      {1., 0., 1., 0.}}}}}, torch::kFloat);
+    ASSERT_TRUE(output.allclose(expected));
+  }
+  {
+    ReflectionPad3d m(ReflectionPad3dOptions({1, 1, 1, 1, 1, 2}));
+    auto input = torch::arange(12, torch::kFloat).reshape({1, 1, 2, 2, 3});
+    auto output = m(input);
+    auto expected = torch::tensor({{{{{10., 9., 10., 11., 10.},
```
Comment (jbschlosser) [inline test/cpp/api/modules.cpp:4018]: Is this correct? I might be reading this wrong, but I see padding=2 for the back and this expected shape only has 1 each on front & back.

File: torch/testing/_internal/common_nn.py [hunk 2]
```diff
@@ -2191,6 +2191,12 @@ def fractional_max_pool3d_test(test_case):
         skip_half=True,
         desc='complex'
     ),
+    dict(
```
Comment (jbschlosser) [inline torch/testing/_internal/common_nn.py:2194]: Should there be an entry that checks complex dtypes like for the 1D and 2D modules?

**Review by jbschlosser — CHANGES_REQUESTED** [review]
Thanks for adding this! You've probably seen this already, but `reflection_pad3d` needs an entry in `tools/autograd/derivatives.yaml` for the backward pass implementation. Looks like some tests are failing, and that'll be at least one reason why. I'll do another pass once that's in :)

File: aten/src/ATen/native/ReflectionPad.cpp [hunk 3]
```diff
@@ -686,5 +976,120 @@ Tensor reflection_pad2d_backward_cpu(
...
+          pad_left, pad_top, pad_front);
+    });
+  }
+}
+
+TORCH_IMPL_FUNC(reflection_pad3d_backward_out_cpu)(const Tensor& grad_output,
+    const Tensor& input,
+    IntArrayRef padding,
+    const Tensor& grad_input) {
+  int64_t pad_left = padding[0];
```
Comment (malfet) [inline aten/src/ATen/native/ReflectionPad.cpp:1037]: Not sure if autogen does it for us, but don't we want to `TORCH_CHECK(padding.size() > 5);`

File: aten/src/ATen/native/ReflectionPad.cpp [hunk 4]
```diff
@@ -686,5 +976,120 @@ Tensor reflection_pad2d_backward_cpu(
...
+  int64_t pad_top = padding[2];
+  int64_t pad_front = padding[4];
+  int64_t dim_w = 3;
+  int64_t dim_h = 2;
+  int64_t dim_d = 1;
+  int64_t dim_plane = 0;
+  bool batch_mode = (input.dim() == 5);
```
Comment (malfet) [inline aten/src/ATen/native/ReflectionPad.cpp:1044]: Please add `TORCH_CHECK(input.dim() >= 4)` and `TORCH_CHECK(grad_output.dim() == input.dim());`
```suggestion
  bool batch_mode = (input.dim() == 5);
  TORCH_CHECK(input.dim() > 3);
  TORCH_CHECK(grad_output.dim() == input.dim());
  TORCH_CHECK(grad_input.dim() == grad_output.dim());
```

File: aten/src/ATen/native/ReflectionPad.cpp (hunk 3 above)
Comment (malfet) [reply aten/src/ATen/native/ReflectionPad.cpp:1037]: @albanD confirmed that autogen is not checking array sizes, so it must be done in a function.

Comment (thomasjpfan) [reply aten/src/ATen/native/ReflectionPad.cpp:1037]: We have a meta function: `TORCH_META_FUNC(reflection_pad3d_backward)` that does the input validation. From my understanding, this means that the implementation does not need to do the validation anymore.

**Comment by ezyang** [comment]
@jbschlosser I'm going to anoint you as shepherd for this patch, thanks!

**Review by jbschlosser — COMMENTED** [review]
Looking pretty good :) Couple comments below for addressing the test failures

**Review by jbschlosser — APPROVED** [review]
LGTM as long as checks pass!
